UI Context stack & 4 character constants

guusw opened the pull request:
Makes the egui context exposed variable a stack.
This way we can have claymore UI host user scripts containing it's own UI.
Additionally I want to make some additional things available within the new `Context` struct, mainly to be able to map a region inside egui back into window coordinates

Kryptos-FR commented on rust/src/shards/gui/mod.rs:
```diff
@@ -66,6 +66,9 @@ impl EguiId {
     }
   }
 }
+pub struct Context<'a> {
```
Missing new line

```suggestion

pub struct Context<'a> {
```

In addition, does it need to be publicly available from outside (e.g. third-party code)?
I would also rename it to `InternalContext` or `EguiContextWrapper` directly here, to not be confused with all the other context types we have 😅.

Kryptos-FR commented on rust/src/shards/gui/mod.rs:
```diff
@@ -79,6 +82,7 @@ struct EguiContext {
   input_translator: egui_gfx::InputTranslator,
 }
 
+
```
Unnecessary new line.
```suggestion
```

Kryptos-FR commented on rust/src/shards/gui/containers/area.rs:
```diff
@@ -6,6 +6,7 @@ use super::Area;
 use crate::shard::Shard;
 use crate::shards::gui::containers::ANCHOR_TYPES;
 use crate::shards::gui::util;
+use crate::shards::gui::util::get_current_context;
```
Just a personal preference, but I like to keep the namespace for those helper functions. Same as with `util::activate_ui_contents(context, input, ui...)` we could have `util::get_current_context(&self.instance)`.

Kryptos-FR commented on rust/src/shards/gui/containers/area.rs:
```diff
@@ -217,11 +218,7 @@ impl Shard for Area {
-        Var::from_object_ptr_mut_ref(*self.instance.get(), &EGUI_CTX_TYPE)?;
-      &*ctx_ptr
-    };
+    let gui_ctx = get_current_context(&self.instance)?;
```
See other comment.

```suggestion
    let gui_ctx = util::get_current_context(&self.instance)?;
```

Kryptos-FR commented on rust/src/shards/gui/containers/panels.rs:
```diff
@@ -8,6 +8,7 @@ use super::RightPanel;
 use super::TopPanel;
 use crate::shard::Shard;
 use crate::shards::gui::util;
+use crate::shards::gui::util::get_current_context;
```
Same.

Kryptos-FR commented on rust/src/shards/gui/containers/window.rs:
```diff
@@ -7,6 +7,7 @@ use crate::shard::Shard;
 use crate::shards::gui::containers::SEQ_OF_WINDOW_FLAGS;
 use crate::shards::gui::containers::WINDOW_FLAGS_TYPE;
 use crate::shards::gui::util;
+use crate::shards::gui::util::get_current_context;
```
Same.

Kryptos-FR commented on rust/src/shards/gui/context.rs:
```diff
@@ -1,6 +1,8 @@
 /* SPDX-License-Identifier: BSD-3-Clause */
 /* Copyright © 2022 Fragcolor Pte. Ltd. */
 
+use super::util::with_object_stack_var;
```
Same.

```suggestion
use super::util;
```

Kryptos-FR reviewed (changes requested):
Overall LGTM. A few suggestions.

guusw commented on rust/src/lib.rs:
```diff
@@ -43,6 +43,14 @@ use crate::types::Var;
 use std::convert::TryInto;
 use std::ffi::CString;
 
+pub const fn fourCharacterCode(val: [u8; 4]) -> i32 {
```
I believe so, it didn't compile without so I think it's similar to constexpr in c++

guusw replied:
yep https://doc.rust-lang.org/reference/const_eval.html

guusw commented on rust/src/shards/gui/util.rs:
```diff
@@ -9,36 +10,56 @@ use crate::types::ExposedTypes;
-) -> Result<Var, &'a str> {
-  // pass the ui parent to the inner shards
-  let var = parents.get();
+use super::EGUI_CTX_TYPE;
```
I'm not gonna change all of them here, since a bunch of other gui files also use `use super::<something>`

guusw commented:
Added a test `src/tests/ui-nested.edn` and also fixed a case where the nested UI would take UI parents from the outer UI (in case of CentralPanel for example)

Also I believe there was a potential leak/corruption issue with the stack variables because they:
- retrieved the seq (by value)
- pushed 
- set it on ParamVar
- run some other code
- pop from the **SAME** seq as in the first step (here might be an issue since the memory could've been reallocated/resized, but it won't be reflected here)
- set it on ParamVar again

Should be fixed by retrieving from the ParamVar again after running the inner code

guusw commented on rust/src/shards/gui/context.rs:
```diff
@@ -174,11 +175,18 @@ impl Shard for EguiContext {
     // Initialize the parents stack in the root UI.
     // Every other UI elements will reference it and push or pop UIs to it.
-    self.parents.set(Seq::new().as_ref().into());
+    if !self.parents.get().is_seq() {
```
I'm not sure if this was an issue but I expect this might clear the stack if some nested UI were to be warmed up from within a (Setup) block in another UI
